src/plugins/action_generate_review_assignment_records/src/generateReviewAssignments.ts

```diff
@@ -1,112 +1,105 @@
-import { Reviewer, ReviewAssignmentObject, AssignmentStatus } from './types'
+import { Reviewer, ReviewAssignmentObject, AssignmentStatus, ApplicationData } from './types'
 import databaseMethods from './databaseMethods'
 
 module.exports['generateReviewAssignments'] = async function (input: any, DBConnect: any) {
   const db = databaseMethods(DBConnect)
 
   console.log('Generating review assignment records...')
   try {
-    const { applicationId, reviewId, templateId, stageId, stageNumber } = input
-    // NB: reviewId comes from record_id on TriggerPayload when triggered
-    // from review table
+    const { applicationId, reviewId } = input
+    // Get template information and current stage for application
+    const applicationData: ApplicationData = await DBConnect.getApplicationData(applicationId)
+    const { templateId, stageNumber, stageId } = applicationData
 
-    const numReviewLevels: number = await DBConnect.getNumReviewLevels(templateId, stageNumber)
+    const numReviewLevels: number =
+      (await DBConnect.getNumReviewLevels(templateId, stageNumber)) || 0
 
-    const currentReviewLevel: number = reviewId
-      ? await DBConnect.getCurrentReviewLevel(reviewId)
-      : 0
-    if (currentReviewLevel === numReviewLevels) {
-      console.log(
-        'Final review level reached for current stage, no later review assignments to generate.'
-      )
-      return {
-        status: 'Success',
-        error_log: '',
-        output: {},
+    // NB: reviewId comes from record_id on TriggerPayload when triggered from review table
+    // For first review assignment
+    if (!reviewId) {
+      console.log('First review assignment on first stage - total levels:', numReviewLevels)
+      if (numReviewLevels === 0) {
+        console.log(
+          'No reviewer with level associated to first stage, no review assignments to generate.'
+        )
+        return {
+          status: 'Success',
+          error_log: '',
+          output: {},
+        }
       }
+      const nextReviewLevel = 1
+      const isLastLevel = nextReviewLevel === numReviewLevels
+      return generateNextReviewAssignments({
+        db,
+        applicationId,
+        templateId,
+        nextReviewLevel,
+        nextStageNumber: stageNumber,
+        nextStageId: stageId,
+        isLastLevel,
+      })
```

Comment:
This block is basically the same as in the `else` block below, so probably want to move it outside the `if/else`.

Reply:
Hmm I'll see if I can do this - was just trying to get this working properly first. Now it's tricky to test on master. But let's see

Reply:
Ah, ok. So this block needs to be separated so it knows to use this `stageNumber` and don't do the checking to check `getReviewStageAndLevel` is in the same stage as application or not. This can be simplified for sure. I just don't want to focus on that, to be able to do other stuff (it already took half day for me to find this solution and test it properly). I think the 2 blocks look similar but the main task is done by the `generateNextReviewAssignments` so that's fine by me.